feat(devtools): add transfer state tab

Avcharov opened the pull request:
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/contributing-docs/commit-message-guidelines.md
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.dev application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
transfer state is visible only on elements tab in ng-state script.

Issue Number: #62237 

## What is the new behavior?
Did workaround with adding of new transfer state tab. Still consider solution with transfer state list in components tab intead of proposed. Let me know which one is prefered.

Demo:
![Click here for Demo Video](https://github.com/user-attachments/assets/91b0f6f1-e395-4935-a246-78d44bf4394d)

<img width="1097" alt="Снимок экрана 2025-07-04 в 10 30 01" src="https://github.com/user-attachments/assets/fb505eb8-3c6e-456c-b512-b63917913b6f" />


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

hawkgs commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/transfer-state/transfer-state.component.scss:
```diff
@@ -0,0 +1,347 @@
+          letter-spacing: 0.5px;
+          color: #000000;
+
+          &.type-string {
+            background: #a6d5a9;
+          }
+
+          &.type-number {
+            background: #c79eff;
+          }
+
+          &.type-boolean {
+            background: #f9c2c5;
+          }
+
+          &.type-object {
+            background: #ffd54f;
+          }
```
We've been trying to limit direct color usage in our styles. You can take a look at `_colors.scss`, there are some `dynamic-` colors that change automatically when the theme is switched. Ultimately, it would be great, if you could reuse the colors from the existing palette, but feel free to add new ones if you need them.

hawkgs commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/transfer-state/transfer-state.component.scss:
```diff
@@ -0,0 +1,347 @@
+              background: #f8f9fa;
+              border: 1px solid #e1e4e8;
+              border-radius: 3px;
+              font-family: 'Monaco', 'Menlo', 'Ubuntu Mono', monospace;
+              font-size: 11px;
+              line-height: 1.4;
```
Same for typography – there are some existing typography styles in `_typography.scss` that you could employ.

hawkgs commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/transfer-state/transfer-state.component.scss:
```diff
@@ -0,0 +1,347 @@
+        }
+
+        td {
+          padding: 12px 16px;
```
Prefer `rem` over `px`.

hawkgs commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/transfer-state/transfer-state.component.html:
```diff
@@ -0,0 +1,132 @@
+      </mat-card-content>
+    </mat-card>
+  } @else if (!hasData()) {
+    <mat-card class="empty-card">
+      <mat-card-header>
+        <mat-card-title>
+          <mat-icon>info</mat-icon>
+          Transfer State is Empty
+        </mat-card-title>
+      </mat-card-header>
```
You can also consider this an answer to your "custom layout" question: We are not completely Material-free, nor do we have plans to become such. For example, we still use `mat-table` and `mat-icon` (to name a few) since they are very handy. However, we try to avoid very Material-design-heavy components since, understandably, we aim to achieve a design language similar to the browsers' DevTools. That being said, components that can be easily recreated with a `div` or two and several lines of CSS, for instance, are preferred.

JeanMeche commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/transfer-state/transfer-state.component.ts:
```diff
@@ -0,0 +1,211 @@
+
+@Component({
+  selector: 'ng-transfer-state',
+  standalone: true,
```
nit: those aren't necessary anymore (since v19). 
```suggestion
```

JeanMeche replied:
can also remove it, thank you.

JeanMeche commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/devtools-tabs.component.html:
```diff
@@ -153,6 +158,14 @@
         >
       </label>
     }
+    <label mat-menu-item disableRipple>
+      <mat-slide-toggle
+        [checked]="transferStateTabEnabled()"
+        (change)="setTransferStateTab($event.checked)"
+      >
+        Show Transfer State Tab</mat-slide-toggle
+      >
+    </label>
```
We should hide this toggle if the API isn't availble.

hawkgs commented on devtools/projects/ng-devtools-backend/src/lib/ng-debug-api/ng-debug-api.ts:
```diff
@@ -81,3 +81,12 @@ export function ngDebugSignalGraphApiIsSupported(): boolean {
+ * Checks if transfer state is available.
+ * Transfer state is only relevant when Angular app uses Server-Side Rendering.
+ */
+export function ngDebugTransferStateApiIsSupported(): boolean {
```
Could you please add a test in `ng-debug-api.spec.ts` to keep things consistent?

hawkgs commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/transfer-state/transfer-state.component.ts:
```diff
@@ -0,0 +1,211 @@
+    try {
+      const str = JSON.stringify(value);
+      const bytes = new Blob([str]).size;
+      if (bytes < 1024) return `${bytes} B`;
+      if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(1)} KB`;
+      return `${(bytes / (1024 * 1024)).toFixed(1)} MB`;
```
Can we export this in a utility function? It seems that `totalSize` uses the same logic so I think it makes sense to reuse this.

hawkgs commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/transfer-state/transfer-state.component.ts:
```diff
@@ -0,0 +1,211 @@
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class TransferStateComponent {
+  private _messageBus = inject(MessageBus) as MessageBus<Events>;
+  private _clipboard = inject(Clipboard);
```
Style: We are no longer prefixing the private members with underscores.

hawkgs commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/transfer-state/transfer-state.component.ts:
```diff
@@ -0,0 +1,211 @@
+    return text.substring(0, this.MAX_DISPLAY_LENGTH) + '...';
+  }
+
+  loadTransferState(): void {
```
`private`?

hawkgs commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/transfer-state/transfer-state.component.scss:
```diff
@@ -0,0 +1,366 @@
+  }
+}
+
+@include theme.dark-theme {
```
Thanks, I appreciate your efforts for converting the colors!

That being said, I don't think you need this dark theme mixin. From what I've checked, the stylesheet completely relies on dynamic colors* which alter themselves during theme change, so you don't really need to do it yourself (e.g. `--color-text` variable gets automatically updated from black to white and vice versa when you switch the theme). 

The `theme.dark-theme` and `theme.light-theme` mixins are generally for special cases.

_* All colors in `_light-colors` and `_dark-colors` mixins in `_colors.scss`._

dgp1130 commented on devtools/projects/ng-devtools-backend/src/lib/client-event-subscribers.ts:
```diff
@@ -661,6 +664,40 @@ const logProvider = (
+    return;
+  }
+
+  const transferStateData = ng.ɵgetTransferState?.(injector) as Record<string, TransferStateValue>;
```
Nit: I think this should be `| undefined` since `ng.ɵgetTransferState` may not exist.

hawkgs replied:
I know that currently if the protocol event data type and `ɵgetTransferState` return type diverge, we will get a compile-time error that they don't sufficiently overlap, but maybe it's safer to drop the assertion altogether?

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/devtools-tabs.component.ts:
```diff
@@ -67,6 +69,7 @@ export class DevToolsTabsComponent {
   readonly routerGraphEnabled = signal(false);
   readonly timingAPIEnabled = signal(false);
   readonly signalGraphEnabled = signal(false);
+  readonly transferStateTabEnabled = signal(false);
```
Follow up: @hawkgs, we should save this value as well in https://github.com/angular/angular/pull/62429 if it lands first.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/transfer-state/transfer-state.component.html:
```diff
@@ -0,0 +1,133 @@
+          <ng-container matColumnDef="size">
+            <th mat-header-cell *matHeaderCellDef>Size</th>
+            <td mat-cell *matCellDef="let item">{{ item.size }}</td>
+          </ng-container>
```
Consider: Should we put a tooltip on the column (maybe with a `?` icon) indicating that this is uncompressed size?

I know on Angular CLI bundle budgets a lot of people get confused whether we're listing disk size or transfer size, so it might be helpful to be explicit.

Actually we do provide an estimated transfer size based on simple formula. @clydin do you remember how that works? Maybe it would be worth including an estimated transfer size here?

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/transfer-state/transfer-state.component.ts:
```diff
@@ -0,0 +1,193 @@
+      });
+    } catch (err) {
+      this.error.set(`Error loading transfer state: ${err}`);
+      this.isLoading.set(false);
```
Consider: Could we derive `isLoading` from whether either `transferStateData` and `error` have a value provided?

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/transfer-state/transfer-state.component.html:
```diff
@@ -0,0 +1,133 @@
+            <th mat-header-cell *matHeaderCellDef>Value</th>
+            <td mat-cell *matCellDef="let item" class="value-cell">
+              <div class="value-container">
+                <pre
```
Consider: Should this include `<code>`? Would that be better for accessibility?

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/transfer-state/transfer-state.component.scss:
```diff
@@ -0,0 +1,289 @@
+            align-items: flex-start;
+            gap: 0.5rem;
+
+            .value-preview {
```
UX: TBH from the screenshot / video I genuinely thought this was a `textarea`.

@hawkgs what do you think about a style like markdown preview of a codeblock (screenshot from VSCode)? Would that be more intuitive?

<img width="681" height="295" alt="Screenshot from 2025-07-14 16-30-42" src="https://github.com/user-attachments/assets/1b313dc4-1539-4b67-b0fb-334d71aee3b8" />

hawkgs replied:
I think that the colors can be changed to match the theme, yes. 

Overall, we can do some minor tweaking here and there but we can introduce these changes in a follow-up PRs so this one isn't blocked.

dgp1130 commented on packages/core/src/render3/util/transfer_state_utils.ts:
```diff
@@ -0,0 +1,38 @@
+    ...result
+  } = transferState;
+
+  return result;
```
Consider: You can drop the `_` destructures if you overwrite in a new object:

```typescript
return {
  ...result,
  [TRANSFER_STATE_TOKEN_ID]: undefined,
  [TRANSFER_STATE_DEFER_BLOCKS_INFO]: undefined,
};
```

You might have to filter out keys with `undefined` values when consuming this. Alternatively, you can just `delete` the keys.

```typescript
const transferState = retrieveTransferredState(doc, appId);
delete transferState[TRANSFER_STATE_TOKEN_ID];
delete transferState[TRANSFER_STATE_DEFER_BLOCKS_INFO];
```

dgp1130 reviewed (approved):
Thanks for putting this together, very cool feature!

No major concerns on my side, just a few minor suggestions and potential UX improvements.

hawkgs commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/transfer-state/transfer-state.component.scss:
```diff
@@ -0,0 +1,289 @@
+@use '../../../styles/theme' as theme;
```
Nit: Looks like a leftover. We don't need this anymore. (The Bazel `deps` can be updated as well).

Avcharov commented:
upd: colors to match theme
<img width="869" height="677" alt="Снимок экрана 2025-07-18 в 07 39 59" src="https://github.com/user-attachments/assets/d84e51b9-354c-4531-bb55-b32290875534" />

Avcharov committed:
feat(devtools): add transfer state tab

Add transfer state tab, which is taking transfer state script by using APP_ID. Created internal api ɵgetTransferState to retrieve transfer state value from app into devtools app.

JeanMeche commented:
@Avcharov I've allowed myself to force push an update following the merge of #62627

AndrewKushnir commented on packages/core/src/render3/util/transfer_state_utils.ts:
```diff
@@ -0,0 +1,35 @@
+  const transferState = retrieveTransferredState(doc, appId);
+
+  // Strip internal keys
+  delete transferState[TRANSFER_STATE_TOKEN_ID];
+  delete transferState[TRANSFER_STATE_DEFER_BLOCKS_INFO];
```
It'd be great to add a test for this change, to make sure we do not "leak" other private fields (we may add more in the future) into the transfer state visualization.

Unrelated (for followup PRs): it'd be great to rename `TRANSFER_STATE_TOKEN_ID` -> `TRANSFER_STATE_HYDRATION_INFO` to better represent the role of the symbol.

AndrewKushnir replied:
cc @JeanMeche

JeanMeche replied:
I've added a bunch of tests 👍

AndrewKushnir reviewed (commented):
Looks great, just left one comment.

AndrewKushnir commented on packages/core/src/hydration/utils.ts:
```diff
@@ -68,6 +68,14 @@ export const NGH_DEFER_BLOCKS_KEY: StateKey<{[key: string]: SerializedDeferBlock
   [key: string]: SerializedDeferBlock;
 }>(TRANSFER_STATE_DEFER_BLOCKS_INFO);
 
+/**
+ * Set of transfer state keys that are used internally by the framework
+ */
+export const internalHydrationTransferStateKeys = [
+  TRANSFER_STATE_TOKEN_ID,
+  TRANSFER_STATE_DEFER_BLOCKS_INFO,
+];
+
```
I'd like to propose a couple of changes:

- Create a function (instead of an array) as mentioned below
- Use the function to filter out such keys from the transfer state (in the logic inside of `getTransferState`)
- For tests, we can use regular hydration specs as a base, just add a few more keys to the transfer state and after that assess whether `getTransferState` returns only public/application keys

```suggestion
/**
 * Checks whether a given key is used by the framework for transferring hydration data.
 */
export function isInternalHydrationTransferStateKey(key: string): boolean {
  return key === TRANSFER_STATE_TOKEN_ID || key === TRANSFER_STATE_DEFER_BLOCKS_INFO;
}
```

JeanMeche replied:
I've extracted the core part of this PR into a separate PR #62722

AndrewKushnir reviewed (commented):
@JeanMeche thanks for the update, I've left a comment with a couple of suggestions.

JeanMeche commented:
Blocked on #62722

Avcharov commented:
@JeanMeche what you think about adding checkbox "Hide internal keys" on top of the page to hide hydration and other internal keys conditionally?

As sometimes maybe that keys can be useful, as @eneajaho [mentioned in this PR](https://github.com/angular/angular/pull/62722#discussion_r2218221682_)

JeanMeche commented:
PR is now unblocked, we need a final approval and we can send it to the merge queue !

dgp1130 reviewed (approved):
This all looks good to me, happy to merge this as is and we can iterate more as we move forward with it.

Thanks @Avcharov for contributing this! It's a very cool addition to Angular DevTools.

thePunderWoman reviewed (approved):
LGTM!

reviewed-for: fw-general

kirjs commented:
This PR was merged into the repository by commit a2f366f629e43647cb54cb8082f0b0ac198fc4ff.

The changes were merged into the following branches: main
